lisa/src/main/java/it/unive/lisa/analysis/heap/TypeBasedHeap.java

```diff
@@ -0,0 +1,182 @@
+package it.unive.lisa.analysis.heap;
+
+import it.unive.lisa.analysis.BaseHeapDomain;
+import it.unive.lisa.analysis.SemanticException;
+import it.unive.lisa.caches.Caches;
+import it.unive.lisa.program.cfg.ProgramPoint;
+import it.unive.lisa.symbolic.SymbolicExpression;
+import it.unive.lisa.symbolic.heap.AccessChild;
+import it.unive.lisa.symbolic.heap.HeapAllocation;
+import it.unive.lisa.symbolic.heap.HeapExpression;
+import it.unive.lisa.symbolic.heap.HeapReference;
+import it.unive.lisa.symbolic.value.HeapIdentifier;
+import it.unive.lisa.symbolic.value.Identifier;
+import it.unive.lisa.symbolic.value.Skip;
+import it.unive.lisa.symbolic.value.ValueExpression;
+import it.unive.lisa.type.Type;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import org.apache.commons.collections.CollectionUtils;
+
+/**
+ * A type-based heap implementation that abstracts heap locations depending on
+ * their types, i.e., all the heap locations with the same type are abstracted
+ * into a single unique identifier.
+ *
+ * @author <a href="mailto:vincenzo.arceri@unive.it">Vincenzo Arceri</a>
+ */
+public class TypeBasedHeap extends BaseHeapDomain<TypeBasedHeap> {
+
+	private static final TypeBasedHeap TOP = new TypeBasedHeap();
+
+	private static final TypeBasedHeap BOTTOM = new TypeBasedHeap();
+
+	private final Collection<ValueExpression> rewritten;
+
+	private static HashSet<String> NAMES = new HashSet<String>();
```

Comment:
Naming conventions: if it's not a constant (no `final`), then the name is camelCase